Add support for access tokens to cli

argerus opened the pull request:
Adds support for access tokens to databroker-cli.

Included are a number of other improvements. Some necessary for adding this support, some not. E.g.

* Add support for command line arguments
* Add initial support for subcommands
* Start moving GRPC related parts into client.rs
* A number of small updates to the interface

SebastianSchildt commented:
I could get this to run
```
cargo run --bin databroker -- --port 55556 --metadata ../data/vss-core/vss_release_3.0.json --jwt-public-key ../kuksa_certificates/jwt/jwt.key.pub
```

But then I guess I met some documentation issues with client

```
     Running `/Users/scs2rng/Documents/Dev/kuksa.val/target/debug/databroker-cli --server 'http://127.0.0.1:55556' --token-file ../kuksa_certificates/jwt/super-admin.json.token`

  ⠀⠀⠀⢀⣤⣶⣾⣿⢸⣿⣿⣷⣶⣤⡀
  ⠀⠀⣴⣿⡿⠋⣿⣿⠀⠀⠀⠈⠙⢿⣿⣦⠀
  ⠀⣾⣿⠋⠀⠀⣿⣿⠀⠀⣶⣿⠀⠀⠙⣿⣷   
  ⣸⣿⠇⠀⠀⠀⣿⣿⠠⣾⡿⠃⠀⠀⠀⠸⣿⣇⠀⠀⣶⠀⣠⡶⠂⠀⣶⠀⠀⢰⡆⠀⢰⡆⢀⣴⠖⠀⢠⡶⠶⠶⡦⠀⠀⠀⣰⣶⡀
  ⣿⣿⠀⠀⠀⠀⠿⢿⣷⣦⡀⠀⠀⠀⠀⠀⣿⣿⠀⠀⣿⢾⣏⠀⠀⠀⣿⠀⠀⢸⡇⠀⢸⡷⣿⡁⠀⠀⠘⠷⠶⠶⣦⠀⠀⢠⡟⠘⣷
  ⢹⣿⡆⠀⠀⠀⣿⣶⠈⢻⣿⡆⠀⠀⠀⢰⣿⡏⠀⠀⠿⠀⠙⠷⠄⠀⠙⠷⠶⠟⠁⠀⠸⠇⠈⠻⠦⠀⠐⠷⠶⠶⠟⠀⠠⠿⠁⠀⠹⠧
  ⠀⢿⣿⣄⠀⠀⣿⣿⠀⠀⠿⣿⠀⠀⣠⣿⡿
  ⠀⠀⠻⣿⣷⡄⣿⣿⠀⠀⠀⢀⣠⣾⣿⠟    databroker-cli                
  ⠀⠀⠀⠈⠛⠇⢿⣿⣿⣿⣿⡿⠿⠛⠁     v0.3.1                        

Successfully connected to http://127.0.0.1:55556/
sdv.databroker.v1 > get Vehicle.Speed 
[get]  OK  
Vehicle.Speed: ( NotAvailable )
sdv.databroker.v1 > feed Vehicle.Speed 200
unknown command
sdv.databroker.v1 > ?
unknown command
sdv.databroker.v1 > set
unknown command
sdv.databroker.v1 > help

  connect [URI]            Connect to server
  get <PATH> [[PATH] ...]  Get signal value(s)
  actuate <PATH> <VALUE>   Actuate signal
  subscribe <QUERY>        Subscribe to signals using a query
  metadata [FILTER]        Fetch metadata. Provide FILTER to view metadata of signals matching filter.
  provide <PATH> <VALUE>   Provide (publish) a signal value
  token <TOKEN>            Use TOKEN as access token
  token-file <FILE>        Use content of FILE as access token
  help                     You're looking at it.
  quit       
sdv.databroker.v1 > provide Vehicle.Speed 200
[provide]  OK  
sdv.databroker.v1 > actuate Vehicle.Body.Trunk.
Front.  Rear.
sdv.databroker.v1 > actuate Vehicle.Body.Trunk.Rear.Is
IsLocked  IsOpen
sdv.databroker.v1 > actuate Vehicle.Body.Trunk.Rear.IsOpen 
Usage: actuate <PATH> <VALUE>
sdv.databroker.v1 > actuate Vehicle.Body.Trunk.Rear.IsOpen true
[actuate]  Unauthenticated  No auth token provided
sdv.databroker.v1 > 
```

It see `sdv.databroker.v1`, I think that is not a good prompt, Because this should not matter when using the client.. if it does, how would I switch to the "new" API?

I would suggest, that `unknown command` command be extended with sth like `see "help" for a list of available commands`

Also, why is "provide" working, but "actuate" not, is that intended?

Generally, I think READMEs for databroker to be extended/updated to mention these features.

Should also give at least a small example how to use it with databroker-cli (with warning that it is not fully implemented and maybe also mentioning kuksa-client (py) not yet supported)

SebastianSchildt commented on kuksa_databroker/databroker-cli/src/client.rs:
```diff
@@ -0,0 +1,221 @@
+use std::collections::HashMap;
```
#486

argerus commented:
> Also, why is "provide" working, but "actuate" not, is that intended?

That's a bug, will push an update.

> It see sdv.databroker.v1, I think that is not a good prompt, Because this should not matter when using the client.. if it does, how would I switch to the "new" API?

The intention is to add support for `kuksa.val.v1` and the ability to switch between them.

> Generally, I think READMEs for databroker to be extended/updated to mention these features.

Agreed

SebastianSchildt reviewed (changes requested):
✅ : Actuate working now

I still think "Unknown command" should give a hint to type help

This documentation 
https://github.com/eclipse/kuksa.val/tree/master/kuksa_databroker
is outdated:

Need to have updated command line aprameters (and may one sentence for each where not obvious) here : https://github.com/eclipse/kuksa.val/tree/master/kuksa_databroker#test-the-databroker

 It should at leat repair the cli examples here https://github.com/eclipse/kuksa.val/tree/master/kuksa_databroker#test-the-databroker

And update the quickstart hee https://github.com/eclipse/kuksa.val/blob/master/doc/quickstart.md
(not adding authorisation just changing "feed->provide" in the examples of databroker-cli

argerus commented:
> I still think "Unknown command" should give a hint to type help

Done

> Need to have updated command line aprameters (and may one sentence for each where not obvious) here : https://github.com/eclipse/kuksa.val/tree/master/kuksa_databroker#test-the-databroker

Updated it, not sure I got all.

> It should at leat repair the cli examples here https://github.com/eclipse/kuksa.val/tree/master/kuksa_databroker#test-the-databroker

Done as well, I think.. (I reverted the command renaming for now)

SebastianSchildt reviewed (changes requested):
I seem unlucky, I am still facing issues:

```
sdv.databroker.v1 > lol
Unknown command. See `help` for a list of available commands.
sdv.databroker.v1 > help

  connect [URI]            Connect to server
  get <PATH> [[PATH] ...]  Get signal value(s)
  set <PATH> <VALUE>       Set actuator signal
  subscribe <QUERY>        Subscribe to signals with QUERY
  feed <PATH> <VALUE>      Publish signal value
  metadata [PATTERN]       Fetch metadata. Provide PATTERN to list metadata of signals matching pattern.
  token <TOKEN>            Use TOKEN as access token
  token-file <FILE>        Use content of FILE as access token
  help                     You're looking at it.
  quit                     Quit

sdv.databroker.v1 > get Vehicle.ADAS.CruiseControl.IsEnabled 
[get]  OK  
Vehicle.ADAS.CruiseControl.IsEnabled: ( NotAvailable )
sdv.databroker.v1 > feed Vehicle.ADAS.CruiseControl.IsEnabled
Usage: feed <PATH> <VALUE>
sdv.databroker.v1 > feed Vehicle.ADAS.CruiseControl.IsEnabled true
No metadata available for Vehicle.ADAS.CruiseControl.IsEnabled. Needed to determine data type for serialization.
sdv.databroker.v1 > feed Vehicle.Speed 200
No metadata available for Vehicle.Speed. Needed to determine data type for serialization.
sdv.databroker.v1 > metadata Vehicle.ADAS.CruiseControl.IsEnabled
[metadata]  OK  
Path                                 Entry type Data type
Vehicle.ADAS.CruiseControl.IsEnabled Actuator   Bool     
sdv.databroker.v1 > 
```

argerus commented:
```
sdv.databroker.v1 > feed Vehicle.Speed 200
No metadata available for Vehicle.Speed. Needed to determine data type for serialization.
```
> I seem unlucky, I am still facing issues:

No, you just put a spotlight on a usability issue (that wasn't there before adding authorization).

I'm assuming you started unauthenticated, which means the metadata couldn't be fetched on startup. The lack of metadata is communicated, but since this behaviour is new it's not obvious why it's missing. It could have been remedied by running `metadata` and trying again.

But I don't see why you should need to run that manually. Pushed an additional commit adding automatic fetching of metadata after setting a new access token.

SebastianSchildt reviewed (approved):
Finally 🌵
